vissl/losses/nnclr_info_nce_loss.py

```diff
@@ -0,0 +1,220 @@
+        self.buffer_params = buffer_params
+        self.dist_rank = get_rank()
+
+        self.prediction_head_params = prediction_head_params
+        self.queue_size = queue_size
+        self._criterion = torch.nn.CrossEntropyLoss()
+
+        # Create the queue
+        self.register_buffer(
+            "queue", torch.randn(queue_size, self.buffer_params.embedding_dim)
+        )
+        self.queue = nn.functional.normalize(self.queue, dim=1)
+
+        self.register_buffer("queue_ptr", torch.zeros(1, dtype=torch.long))
+        self.initialized = False
+
+        # Will be set by NNCLRHook
+        self.prediction_head = None
+        self.preds = None
+
+    @torch.no_grad()
+    def _dequeue_and_enqueue(self, keys: torch.Tensor):
+        # gather keys before updating queue
+        keys = concat_all_gather(keys)
+        batch_size = keys.shape[0]
+
+        ptr = int(self.queue_ptr)
+
+        if ptr + batch_size > self.queue_size:
+            next_ptr = (ptr + batch_size) % self.queue_size
+
+            # replace the keys at ptr (dequeue and enqueue)
+            self.queue[ptr:] = keys[next_ptr::]
+            self.queue[:next_ptr] = keys[:next_ptr]
+
+            ptr = next_ptr  # move pointer
+        else:
+            # replace the keys at ptr (dequeue and enqueue)
+            self.queue[ptr : ptr + batch_size] = keys
+            ptr = (ptr + batch_size) % self.queue_size  # move pointer
+
+        self.queue_ptr[0] = ptr
+
+    def forward(self, embedding: torch.Tensor):
+        """
+        Calculate the loss. Operates on embeddings tensor.
+        """
+        assert embedding.ndim == 2
+        assert embedding.shape[1] == int(self.buffer_params.embedding_dim)
+
+        batch_size = embedding.shape[0]
+        assert (
+            batch_size % self.num_pos == 0
+        ), "Batch size should be divisible by num_pos"
+
+        if not self.initialized:
+            self.queue = self.queue.to(embedding.device)
+            self.initialized = True
+
+        if is_distributed_training_run():
+            labels = torch.arange(batch_size // self.num_pos) + self.dist_rank * (
+                batch_size // self.num_pos
+            )
+        else:
+            labels = torch.arange(batch_size // self.num_pos)
+
+        labels = labels.to(embedding.device)
+
+        with torch.no_grad():
+            nearest_neighbors = self.queue[(embedding @ self.queue.T).argmax(1)]
+
+        nn_a, nn_b = torch.split(
+            nearest_neighbors, split_size_or_sections=batch_size // self.num_pos, dim=0,
+        )
+
+        predictions_a, predictions_b = torch.split(
+            self.preds, split_size_or_sections=batch_size // self.num_pos, dim=0,
+        )
+
+        logits_ab = (
+            torch.matmul(nn_a, gather_from_all(predictions_b).T) / self.temperature
+        )
+        logits_ba = (
+            torch.matmul(nn_b, gather_from_all(predictions_a).T) / self.temperature
+        )
+
+        loss = (
```

Comment:
@prigoyal,
For now I only use half of the negatives as discussed here: https://github.com/facebookresearch/vissl/issues/316#issuecomment-836708548
I'm running a few tests to see how much it affects performance.